alf/networks/s5/utils.py

```diff
@@ -0,0 +1,311 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import torch
+import triton
+import triton.language as tl
+
+
+def make_HiPPO(N):
+    """ Create a HiPPO-LegS matrix.
+        From https://github.com/srush/annotated-s4/blob/main/s4/s4.py
+        Args:
+            N (int32): state size
+        Returns:
+            N x N HiPPO LegS matrix
+    """
+    P = np.sqrt(1 + 2 * np.arange(N))
+    A = P[:, np.newaxis] * P[np.newaxis, :]
+    A = np.tril(A) - np.diag(np.arange(N))
+    return -A
+
+
+def make_NPLR_HiPPO(N):
+    """
+    Makes components needed for NPLR representation of HiPPO-LegS
+     From https://github.com/srush/annotated-s4/blob/main/s4/s4.py
+    Args:
+        N (int32): state size
+
+    Returns:
+        N x N HiPPO LegS matrix, low-rank factor P, HiPPO input matrix B
+
+    """
+    # Make -HiPPO
+    hippo = make_HiPPO(N)
+
+    # Add in a rank 1 term. Makes it Normal.
+    P = np.sqrt(np.arange(N) + 0.5)
+
+    # HiPPO also specifies the B matrix
+    B = np.sqrt(2 * np.arange(N) + 1.0)
+    return hippo, P, B
+
+
+def make_DPLR_HiPPO(N):
+    """
+    Makes components needed for DPLR representation of HiPPO-LegS
+     From https://github.com/srush/annotated-s4/blob/main/s4/s4.py
+    Note, we will only use the diagonal part
+    Args:
+        N:
+
+    Returns:
+        eigenvalues Lambda, low-rank term P, conjugated HiPPO input matrix B,
+        eigenvectors V, HiPPO B pre-conjugation
+
+    """
+    A, P, B = make_NPLR_HiPPO(N)
+
+    S = A + P[:, np.newaxis] * P[np.newaxis, :]
+
+    S_diag = np.diagonal(S)
+    Lambda_real = np.mean(S_diag) * np.ones_like(S_diag)
+
+    # Diagonalize S to V \Lambda V^*
+    Lambda_imag, V = np.linalg.eigh(S * -1j)
+
+    P = V.conj().T @ P
+    B_orig = B
+    B = V.conj().T @ B
+    return Lambda_real + 1j * Lambda_imag, P, B, V, B_orig
+
+
+@triton.jit
+def diag_ssm_forward_kernel(s_ptr, x_ptr, lambda_ptr, y_ptr, length,
+                            batch_size, dim, BLOCK_SIZE: tl.constexpr):
+    col_idx = tl.program_id(0) * BLOCK_SIZE
+    col_offsets = col_idx + tl.arange(0, BLOCK_SIZE)
+    mask = col_offsets < batch_size * dim
+    s = tl.load(s_ptr + col_offsets, mask=mask, other=0)
+    Lambda = tl.load(lambda_ptr + col_offsets % dim, mask=mask, other=0)
+    for t in range(length):
+        offsets = t * batch_size * dim + col_offsets
+        x = tl.load(x_ptr + offsets, mask=mask, other=0)
+        s = s * Lambda + x
+        tl.store(y_ptr + offsets, s, mask=mask)
+
+
+@triton.jit
+def diag_ssm_backward_kernel(s_ptr, lambda_ptr, y_ptr, grad_s_ptr, grad_x_ptr,
```

Comment:
Mathematically working out the backward rules:

![triton_diag_ssm_backward_deduction](https://github.com/HorizonRobotics/alf/assets/1111035/4595dc36-fb21-41f8-bdb6-9fb79df67d52)

which is consistent with the implementation here, assuming we are looking at a portion of vectors with `offsets` as the indices.